### DISCUSSION TITLE
How to support Inlay Hints in Rust

### DISCUSSION THREAD

**Discussion by e190** [joaotavora/eglot#963] [Ideas]
https://rust-analyzer.github.io/manual.html#inlay-hints
![图片](https://user-images.githubusercontent.com/39932163/169802852-a5506448-b123-4f28-a522-ea50353af70d.png)

And Lsp-mode also has this function
https://emacs-lsp.github.io/lsp-mode/page/lsp-rust-analyzer/#inlay-hints

**Comment by joaotavora** [comment]
We need a PR for this.  A link to the protocol's standard would be a good start, and if it works like I guess it does (like diagnostics) I suppose it should be done via a special type of Flymake diagnostic with certain display properties.

**Comment by gmacon** [comment]
It looks like inlay hints were added in LSP 3.17, which was released a couple weeks back; they are described in [Inlay Hint Request](https://microsoft.github.io/language-server-protocol/specifications/lsp/3.17/specification/#textDocument_inlayHint) *et seq*.

**Reply by joaotavora** [reply]
Thanks.  I'm interested in this feature.  Here are some questions to get started:

* What other servers besides `rust-analyzer` support this?  It seems `clangd` also supports them (https://reviews.llvm.org/D116713), but the more the merrier...

* When should inlay hints be requested? On demand, automatically?

* For what region should inlay hints be requested.  All the buffer? Current function?  Arbitrary region? All of the above?

* How are inlay hints to be represented? Using an overlay is an obvious choice, but we could also opt to hide the inlay hints and only show them when the cursor hover the thing to be hinted to.  However, this might be confused with symbol-specific documentation.  So this requires some thought.

My gut feeling is that a special Flymake backend could be devised for this.  It would avoid re-coding much of the UX, which could be similar.  This is if we see the inlay hints as annotations, which I think is a decent enough model.  They are clearly not "diagnostics", but share many common properties.

**Reply by nemethf** [reply]
For what it worth, I think it would annoy me to see the hints all the time.  A keystroke to show them temporarily (until the next keystroke?) in all the visible buffers would probably more useful. Additionally, it would probably be sometimes useful to put a specific hint to the kill-ring.

(I implemented a feature in eglot-x that inserted the hint under point into the buffer. It relied on the not-yet-standardized version of inlay hints of rust-analyzer.   But I haven't used this command enough to say whether it is useful.  Probably it isn't.)

**Reply by joaotavora** [reply]
> For what it worth, I think it would annoy me to see the hints all the time. 

I think I agree, yes.  I was thinking that that behaviour_could_ be configured, for those who like it.  But I would not make it the priority. The default version would be a hoverable symbol that shows something in the echo area when you _mouse hover_ over it.  Much like Flymake's default behaviour.

The big question here is when to request them.  It would seem that these hints are particularly useful when browsing code in strongly typed languages, when there's not a lot of typing going on.  So I was thinking the "current" function as the range (more or less easily gotten in most well-behaved modes with `beginning-of-defun` -> `end-of-defun`)

The keystroke to show them temporarily is also an interesting idea.  But request _and_ show them? Or just "reveal them"?

And you bring up the invalidation of hints.  I think we have to invalidate aggresively.  I fI have:

```
int32_t x;

blablabla();

x [inlay-hint: int32_t]+= y + 2;   
```

i can't risk having the stale inlay-hint `int32_t` remain if I decide I change the type to `double`

**Reply by skrytebane** [reply]
I think the way it's done in VS Code with `editor.inlayHints.enabled` set to `offUnlessPressed` is quite nice. That way, it only shows them for as long as you hold **Ctrl+Alt**. And as @nemethf says, it would be annoying to always show them. (I guess "holding Ctrl+Alt" doesn't quite fit with how it's done in Emacs, but a simple keybinding that shows all inlay hints and removes them on the next keypress, maybe?)

**Reply by joaotavora** [reply]
That's a good idea @skrytebane, I can imagine it being a very nice interface, which is really the biggest challenge here.  But in Emacs, I'm not sure how one does that with keybindings.  Can you start an Emacs bug report for this feature?  I'd like to continue discussing there.

**Reply by skrytebane** [reply]
I'm guessing it's not possible at the moment, and if it was, it's probably not possible when running Emacs in a console. 🤷🏻 

I just mentioned the keybinding used by VS Code. I think an Emacs-friendly keybinding could also be used, my point was just that it's nice to have a simple keybinding that shows all inlay hints until you press something else. Like how `isearch` shows all matches until you quit it.

**Reply by joaotavora** [reply]
> I'm guessing it's not possible at the moment, and if it was, it's probably not possible when running Emacs in a console. 🤷🏻

Very true.  Well, pity. The temporary enable will have to be some temporary mode mechanism.

**Comment by ShadowMitia** [comment]
Hello there!

I've been working on adding support for inlay hints in eglot, I have a WIP up on my repo here: https://github.com/ShadowMitia/eglot

I've just found this discussion, I'll try and apply some of the ideas here, but please let me know what you think, or what I should change. I've based this on the [code lens PR](https://github.com/joaotavora/eglot/pull/71/files), I figured the implementation would be close (I actually thought inlay hints and code lens were the same feature at first). 

The implementation is still incomplete, there are bits of the standard I haven't got round to yet. Unless you want me to open PR now, I'll wait until the code is a bit more solidified.

If you want to check my code, just run M-x eglot-inlay-hint. To update rerun the command again to see the hints update.

**Reply by nemethf** [reply]
When you're ready, it is better to send the PR as a bug report to Emacs.  Also the diff would be easier to read without the unnecessary white-space changes.  Most probably this will be a legally significant contribution, so you might like to start the copyright assessment process now if you haven't completed your paperwork.  Sometimes the process takes months.

**Comment by dchinmay2** [comment]
I have submitted an improved version of @ShadowMitia 's  patch here: https://mail.gnu.org/archive/html/bug-gnu-emacs/2023-02/msg00760.html

**Reply by dchinmay2** [reply]
@joaotavora 
> The default version would be a hoverable symbol that shows something in the echo area when you mouse hover over it.

Having to do something defeats the purpose IMO: you can already see the type when you move your cursor to a variable (lsp "hover") and parameter names when you move your cursor to a function call. Inlay hints are for the whole buffer.

"hover" showing types:
![image](https://user-images.githubusercontent.com/36493671/218262636-9ced87a8-c6aa-446a-946f-126a95593366.png)


"hover" showing parameter names:
![image](https://user-images.githubusercontent.com/36493671/218262739-268a1a5d-df5b-4c53-88a4-b8a1535b5e04.png)

**Reply by joaotavora** [reply]
Let's continue in the bug tracker discussion initiated by your bug report.

As far as I understand, inlay hints will show the type of many visible expressions at once, so it's different from the "hover" scenario.
